tests/unit/train/test_fp32_master_optimizer.py

```
@@ -0,0 +1,280 @@
+
+
+def test_legacy_state_dict_mismatch_raises_actionable_error():
+    model = nn.ModuleDict(
+        {"a": nn.Linear(4, 4, bias=False), "b": nn.Linear(4, 4, bias=False)}
+    ).to(torch.bfloat16)
+    bare_opt = adamw_factory(list(model.named_parameters()))
+    legacy_state = bare_opt.state_dict()
+
+    model["b"].weight.requires_grad_(False)  # wrapper keeps only 1 of 2 params
+    wrapper = FP32MasterOptimizer(list(model.named_parameters()), adamw_factory)
+    with pytest.raises(ValueError, match="optimizer_state_dict.pt"):
```

Ruff (0.15.20)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)